src/seaofnodes/seaOfNodes__Type.ml

```diff
@@ -0,0 +1,208 @@
+  val get_id : t -> int
+end
+module Identified (M : sig
+    type t
+    val hash : t -> int
+    val equal : t -> t -> bool
+  end) : IDENTIFIED with type elt = M.t = struct
+  type t = int * M.t
+  type elt = M.t
+  let hash (_, x) = M.hash x
+  let equal (_, x) (_, y) = M.equal x y
+
+  let get_value (_, x) = x
+  let get_id (id, _) = id
+end
+
+
+module Binop : sig
+  type t = Add
+  val hash : t -> int
+end = struct
+  type t = Add
+  let hash _ = 1
+end
+
+
+module rec JumpSet : GenericSet.GenericSetSig = GenericSet.Make(Jump)
+and BranchSet : GenericSet.GenericSetSig = GenericSet.Make(Branch)
+
+and Data : sig
+  (** It is possible to pattern match elements of type t, *)
+  (** but you must use the smart constructors to build values *)
+  type t = private Const of int | BinOp of {op: Binop.t; operand1: t; operand2: t} | Phi of Phi.t
+
+  val hash : t -> int
+  val equal : t -> t -> bool
+
+  val pp_dot : out_channel -> t -> unit
+
+  (** Constructors *)
+  val const : int -> t
+  val binop : Binop.t -> t -> t -> t
+  val phi : Phi.t -> t
+
+  (** Unique id of the node *)
+  val get_id : t -> int
+end = struct
+  type t = Const of int | BinOp of {op: Binop.t; operand1: t; operand2: t} | Phi of Phi.t
+
+  let hash = function
+    | Const n -> n
+    | BinOp {op; operand1; operand2} ->
+      Binop.hash op lxor Data.hash operand1 lxor Data.hash operand2
+    | Phi phi -> Phi.hash phi
+
+  let equal t t' =
+    match t, t' with
+    | Const n, Const n' -> n = n'
+    | BinOp {op; operand1; operand2}, BinOp {op = op'; operand1 = operand1'; operand2 = operand2'} ->
+      (* We assume that binops and datas are hashconsed and use the physical equality *)
+      op == op' && operand1 == operand1' && operand2 == operand2'
+    | Phi phi, Phi phi' -> phi == phi'
+    | _ -> false
+
+
+  let pp fmt data = Printf.fprintf fmt "node_%d" (Hashtbl.hash data)
+
+  let rec pp_dot fmt data =
+    match data with
+    | Const n ->
+      Printf.fprintf fmt "%a [ label = \"Const %d\" ]\n" pp data n
+    | BinOp {op= Binop.Add; operand1= op1; operand2= op2} ->
+      Printf.fprintf fmt "%a [ label = \"Add\" ]\n" pp data ;
+      Printf.fprintf fmt "%a -> %a\n" pp data pp op1 ;
+      Printf.fprintf fmt "%a -> %a\n" pp data pp op2 ;
+      pp_dot fmt op1 ;
+      pp_dot fmt op2
+    | _ ->
+      ()
+
+  (* Data.t with tags *)
+  module Id = Identified(Data)
+  (* Weak set to remember previously created values *)
+  module W = Weak.Make(Id)
+
+  let nodes = W.create 5003
+  let cpt = ref 1
+
+  (** if d is on the set, we return the previously constructed element, *)
+  (** otherwise we add it to the set and increment cpt *)
+  let add_or_find d =
+    let n0 = (!cpt, d) in
+    let n = W.merge nodes n0 in
+    if n == n0 then incr cpt;
+    Id.get_value n
+
+  let const n =
+    add_or_find @@ Const n
+
+  let binop op operand1 operand2 =
+    add_or_find @@ BinOp {op; operand1; operand2}
+
+  let phi phi =
+    add_or_find @@ Phi phi
+
+  let get_id data =
```

Comment:
I'm unsure about this design. If this operation is called many times, we could just stick the int into the data itself, instead how having to query the HashTable. 
```suggestion
  type term = private Const of int | BinOp of {op: Binop.t; operand1: t; operand2: t} | Phi of Phi.t
  and t = { unique: int; term: term }
```

Reply:
We considered this option, but it makes pattern matching bloated, as we have to ignore the tag everywhere.
The idea of the identified value is to hide the unique tag.
Indeed, this forces us to request the hashset to retrieve the identifier of a value, but we considered this a good trade-off.